protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,258 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\services;
+
+use humhub\commands\MigrateController;
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+
+    protected Module $module;
+    protected ?string $path;
+
+    /**
+     * @param Module $module
+     * @param array $config = ['path' => string|null]
+     */
+    public function __construct(Module $module, array $config = [])
+    {
+        $this->module = $module;
+
+        $path = $config['path'] ?? null;
+        unset($config['path']);
+
+        if ($path === null) {
+            $path = $module->basePath . '/migrations';
+
+            if (!is_dir(realpath($path))) {
+                Yii::debug("Module has no migrations directory.", $this->module->id);
+                $path = null;
+            }
+        } elseif (false === realpath($path)) {
+            $message = sprintf(
+                "Migration path '%s' for module '%s' is invalid!",
+                $path,
+                $this->module->getName()
+            );
+            Yii::error($message, $this->module->id);
+            throw new \RuntimeException($message);
+        }
+
+        $this->path = $path;
+
+        parent::__construct($config);
+    }
+
+    public function getModule(): Module
+    {
+        return $this->module;
+    }
+
+    public function getPath(): ?string
+    {
+        return $this->path;
+    }
+
+    public function migrateUp(string $action): ?bool
+    {
+        $result = new MigrationEvent([
+            'path' => $this->path,
+            'action' => $action,
+            'migration' => 'up',
+            'deactivateOnError' => true
```

Comment:
- Not sure the `MigrationEvent` should handle deactivation?
- Do we really want to delete? Currently deactive === data loses

Reply:
Are you happy to keep this like that for now? But we might address it again in the PR (which I have already ready) that refactors module update and migration

Reply:
Ok, sounds fair.

We use the events for the before/after hooks.